### PR TITLE
[flutter_tts] Manage a speaking state on the state changed callback

### LINKED ISSUE #87: [flutter_tts] PlatformException (Invalid Operation) when play has been canceled
How to reproduce:

1. Launch the example app. (I used a TV 6.0 emulator.)
2. Enter some text in the text field and click PLAY, and then click STOP or PAUSE while the audio is playing.
3. Click PLAY again and see this error message:

```
I/ConsoleMessage(11739): flutter: Playing
I/ConsoleMessage(11739): flutter: Cancel
E/ConsoleMessage(11739): [ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: PlatformException(Invalid Operation, Invalid Operation, null, null)
E/ConsoleMessage(11739): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:581:7)
E/ConsoleMessage(11739): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
E/ConsoleMessage(11739): <asynchronous suspension>
E/ConsoleMessage(11739): #2      _MyAppState._speak (package:flutter_tts_tizen_example/main.dart:90:9)
E/ConsoleMessage(11739): <asynchronous suspension>
```

cc @rookiejava @flutter-tizen/maintainers

### PR DESCRIPTION
**Pull request by bbrto21**
* Fix #87

Signed-off-by: Boram Bae <boram21.bae@samsung.com>

### CODE REVIEW

**Comment by swift-kim** [comment]
Probably not caused by this commit but I'm getting this error when trying to play any text on wearable 5.5 emulator:

```
E/ConsoleMessage(30784): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: PlatformException(-49348607, Failed to speak(tts_add_text)., null, null)
E/ConsoleMessage(30784): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:597:7)
E/ConsoleMessage(30784): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
E/ConsoleMessage(30784): <asynchronous suspension>
E/ConsoleMessage(30784): #2      _MyAppState._speak (package:flutter_tts_tizen_example/main.dart:90:9)
E/ConsoleMessage(30784): <asynchronous suspension>
E/ConsoleMessage(30784):
```

**Comment by swift-kim** [comment]
Another warning on TM1 and R800:

```
I/ConsoleMessage(23522): flutter: Playing
I/ConsoleMessage(23522): flutter: Cancel
E/ConsoleMessage(23522): Warning: Failed to respond to a message. This is a memory leak.
I/ConsoleMessage(23522): flutter: Playing
I/ConsoleMessage(23522): flutter: Cancel
E/ConsoleMessage(23522): Warning: Failed to respond to a message. This is a memory leak.
E/ConsoleMessage(23522):
I/ConsoleMessage(23522): flutter: Playing
I/ConsoleMessage(23522): flutter: Cancel
```

**Comment by bbrto21** [comment]
> Probably not caused by this commit but I'm getting this error when trying to play any text on wearable 5.5 emulator:
> 
> ```
> E/ConsoleMessage(30784): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: PlatformException(-49348607, Failed to speak(tts_add_text)., null, null)
> E/ConsoleMessage(30784): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:597:7)
> E/ConsoleMessage(30784): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
> E/ConsoleMessage(30784): <asynchronous suspension>
> E/ConsoleMessage(30784): #2      _MyAppState._speak (package:flutter_tts_tizen_example/main.dart:90:9)
> E/ConsoleMessage(30784): <asynchronous suspension>
> E/ConsoleMessage(30784):
> ```

**Comment by bbrto21** [comment]
> Probably not caused by this commit but I'm getting this error when trying to play any text on wearable 5.5 emulator:
> 
> ```
> E/ConsoleMessage(30784): [ERROR:flutter/lib/ui/ui_dart_state.cc(199)] Unhandled Exception: PlatformException(-49348607, Failed to speak(tts_add_text)., null, null)
> E/ConsoleMessage(30784): #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:597:7)
> E/ConsoleMessage(30784): #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
> E/ConsoleMessage(30784): <asynchronous suspension>
> E/ConsoleMessage(30784): #2      _MyAppState._speak (package:flutter_tts_tizen_example/main.dart:90:9)
> E/ConsoleMessage(30784): <asynchronous suspension>
> E/ConsoleMessage(30784):
> ```

I was only getting this error when I pressed the play button while speaking, and I thought this error is reasonable.

**Comment by bbrto21** [comment]
> Another warning on TM1 and R800:
> 
> ```
> I/ConsoleMessage(23522): flutter: Playing
> I/ConsoleMessage(23522): flutter: Cancel
> E/ConsoleMessage(23522): Warning: Failed to respond to a message. This is a memory leak.
> I/ConsoleMessage(23522): flutter: Playing
> I/ConsoleMessage(23522): flutter: Cancel
> E/ConsoleMessage(23522): Warning: Failed to respond to a message. This is a memory leak.
> E/ConsoleMessage(23522):
> I/ConsoleMessage(23522): flutter: Playing
> I/ConsoleMessage(23522): flutter: Cancel
> ```

I can't reproduce, please tell me how to reproduce.

File: packages/flutter_tts/tizen/src/flutter_tts_tizen_plugin.cc [hunk 1]
```diff
@@ -241,14 +240,17 @@ class FlutterTtsTizenPlugin : public flutter::Plugin {
...
         std::make_unique<flutter::EncodableValue>(true);
     if (current == TTS_STATE_PLAYING) {
+      plugin->speaking_ = true;
       if (previous == TTS_STATE_READY) {
         plugin->channel_->InvokeMethod("speak.onStart", std::move(args));
       } else if (previous == TTS_STATE_PAUSED) {
         plugin->channel_->InvokeMethod("speak.onContinue", std::move(args));
       }
     } else if (current == TTS_STATE_PAUSED) {
+      plugin->speaking_ = false;
```
Comment (swift-kim) [inline packages/flutter_tts/tizen/src/flutter_tts_tizen_plugin.cc:250]: Maybe we need something like this here and in the next:

```cpp
if (previous == TTS_STATE_PLAYING && plugin->await_speak_completion_) {
  plugin->result_->Success();
}
```

Comment (bbrto21) [reply packages/flutter_tts/tizen/src/flutter_tts_tizen_plugin.cc]: Thank you for suggestion, please review again

File: packages/flutter_tts/tizen/src/flutter_tts_tizen_plugin.cc [hunk 2]
```diff
@@ -241,15 +247,18 @@ class FlutterTtsTizenPlugin : public flutter::Plugin {
...
       } else if (previous == TTS_STATE_PAUSED) {
         plugin->channel_->InvokeMethod("speak.onContinue", std::move(args));
       }
     } else if (current == TTS_STATE_PAUSED) {
+      plugin->EndSpeaking();
       plugin->channel_->InvokeMethod("speak.onPause", std::move(args));
     } else if (current == TTS_STATE_READY) {
       if (previous == TTS_STATE_PLAYING || previous == TTS_STATE_PAUSED) {
+        plugin->EndSpeaking();
         plugin->channel_->InvokeMethod("speak.onCancel", std::move(args));
```
Comment (swift-kim) [inline packages/flutter_tts/tizen/src/flutter_tts_tizen_plugin.cc:262]: Known issue: The "speak.onCancel" method is always called even when the utterance completes successfully.

**Review by swift-kim — APPROVED** [review]

**Review by bwikbs — APPROVED** [review]
Thanks for your working! :smile:

**Pull request merged**